File: src/test/scala/com/exasol/cloudetl/it/scriptclasses/ExportParallelismIT.scala
package com.exasol.cloudetl.it.scriptclasses

import java.util.List
import java.util.stream.Stream
import com.exasol.cloudetl.BaseS3IntegrationTest
import com.exasol.dbbuilder.dialects.Table
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.lessThanOrEqualTo
import org.scalatest.{BeforeAndAfterEach, Ignore}

@Ignore
class ExportParallelismIT extends BaseS3IntegrationTest with BeforeAndAfterEach {
  val SCHEMA_NAME: String = "EXPORT_PARALLELISM"
  val bucketName: String = "export-parallelism"
  var exportTable: Table = _

  override final def beforeAll(): Unit = {
    super.beforeAll()
    prepareExasolDatabase(SCHEMA_NAME)
    createS3ConnectionObject()
    createBucket(bucketName)
    exportTable = schema
      .createTable("T1_PARALLELISM", "C1", "DECIMAL(18,0)")
      .bulkInsert(Stream.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).map(List.of(_)))
  }

  override final def afterEach(): Unit =
    deleteBucketObjects(bucketName)

  test("exports using user provided parallelism") {
    exportData("PARALLELISM = 'iproc()'")
    assertThat(listObjects(bucketName).size(), equalTo(1))
  }

  test("exports using user provided parallelism as iproc() combination") {
    exportData("PARALLELISM = 'iproc(), mod(rownum,3)'")
    assertThat(listObjects(bucketName).size(), equalTo(3))
  }

  test("exports using auto parallelism") {
    exportData("")
    assertThat(listObjects(bucketName).size(), equalTo(2))
  }

  test("exports using auto parallelism with udf memory parameter") {
    exportData("UDF_MEMORY = '250'")
    // export does not have to use all instances, therefore comparing with lessThanOrEqualTo
    assertThat(listObjects(bucketName).size(): Integer, lessThanOrEqualTo(Integer.valueOf(5)))
  }

  def exportData(parallelism: String): Unit =
    executeStmt(
      s"""|EXPORT ${exportTable.getFullyQualifiedName()}
          |INTO SCRIPT $SCHEMA_NAME.EXPORT_PATH WITH
          |BUCKET_PATH     = 's3a://$bucketName/'
          |DATA_FORMAT     = 'PARQUET'
          |S3_ENDPOINT     = '$s3Endpoint'
          |CONNECTION_NAME = 'S3_CONNECTION'
          |$parallelism
          |;
      """.stripMargin
    )

}

File: src/test/java/com/exasol/cloudetl/it/scriptclasses/ExportParallelismIT.java
package com.exasol.cloudetl.it.scriptclasses;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

import java.util.List;
import java.util.stream.Stream;

import org.junit.jupiter.api.*;

import com.exasol.cloudetl.BaseS3IntegrationTest;
import com.exasol.dbbuilder.dialects.Table;

class ExportParallelismIT extends BaseS3IntegrationTest {
    private static final String SCHEMA_NAME = "EXPORT_PARALLELISM";
    private static final String BUCKET_NAME = "export-parallelism";
    private Table exportTable;

    @BeforeAll
    void beforeAll() {
        prepareExasolDatabase(SCHEMA_NAME);
        createS3ConnectionObject();
        createBucket(BUCKET_NAME);
        this.exportTable = this.schema.createTable("T1_PARALLELISM", "C1", "DECIMAL(18,0)")
                .bulkInsert(Stream.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).map(List::of));
    }

    @AfterEach
    void afterEach() {
        deleteBucketObjects(BUCKET_NAME);
    }

    @Test
    void exportsUsingUserProvidedParallelism() {
        exportData("PARALLELISM = 'iproc()'");
        assertThat(listObjects(BUCKET_NAME).size(), equalTo(1));
    }

    @Test
    void exportsUsingUserProvidedParallelismAsIprocCombination() {
        exportData("PARALLELISM = 'iproc(), mod(rownum,3)'");
        assertThat(listObjects(BUCKET_NAME).size(), equalTo(3));
    }

    @Test
    void exportsUsingAutoParallelism() {
        exportData("");
        assertThat(listObjects(BUCKET_NAME).size(), equalTo(2));
    }

    @Test
    void exportsUsingAutoParallelismWithUdfMemoryParameter() {
        exportData("UDF_MEMORY = '250'");
        assertThat(Integer.valueOf(listObjects(BUCKET_NAME).size()), lessThanOrEqualTo(Integer.valueOf(5)));
    }

    private void exportData(final String parallelism) {
        executeStmt(String.format("EXPORT %s%n"
                + "INTO SCRIPT %s.EXPORT_PATH WITH%n"
                + "BUCKET_PATH     = 's3a://%s/'%n"
                + "DATA_FORMAT     = 'PARQUET'%n"
                + "S3_ENDPOINT     = '%s'%n"
                + "CONNECTION_NAME = 'S3_CONNECTION'%n"
                + "%s%n"
                + ";%n", this.exportTable.getFullyQualifiedName(), SCHEMA_NAME, BUCKET_NAME, this.s3Endpoint,
                parallelism));
    }
}
